cip/CIP2016-06-14-Define-comparability-and-equality-as-well-as-orderability-and-equivalence.adoc

```diff
@@ -0,0 +1,377 @@
+
+Orderability and equivalence produce <<missing-null,"missing" `null` values>>.
+
+[[orderability-def,orderability]]
+==== Orderability ====
+
+[[global-sort-order,global sort order]]
+We propose that orderability be defined between any pair of values, where the result is always `true` or `false`.
+
+To accomplish this, there must be a pre-determined order of types and each value must fall under exactly one disjoint leaf type in this order. We propose to use the following ascending global sort order of disjoint types:
+
+* `MAP` types
+** <<regular-map,Regular map>>
+** `NODE`
+** `RELATIONSHIP`
+* `LIST OF ANY?`
+* `PATH`
+* `STRING`
+* `BOOLEAN`
+* `NUMBER`
+** `NaN` is treated as the largest number in orderability only (i.e. it is put after positive infinity)
+* `VOID` (i.e. the type of `null`)
+
+To give a concrete example, under this global sort order all nodes come before all strings.
+
+Between values of the same type in the global sort order, orderability defers to comparability except that equality is overriden by equivalence as described below.
+
+The accompanying descending global sort order is the same order in reverse (i.e. it runs from `VOID` to `MAP`).
+
+[[equivalence-def,equivalence]]
+==== Equivalence ====
+
+Equivalence remains mostly unchanged but now can be defined succinctly as being identical to equality except that:
+
+* Any two `null` values are equivalent (both directly or inside nested structures).
+* Any two `NaN` values are equivalent (both directly or inside nested structures).
+* However, `null` and `NaN` values are not equivalent (both directly or inside nested structures).
+
+Equivalence is reflexive for all values.
+
+[[aggregation]]
+=== Aggregation ===
+
+Generally, an aggregation function call `aggr(expr)` processes all matching rows for each aggregation key found in an incoming record (keys are compared using equivalence, as noted above).
+For a fixed aggregation key and each matching record, `expr` is evaluated to a value.
+This yields a list of candidate values.
+Generally, the order of candidate values is unspecified.
+However, if the aggregation happens in a projection with an associated `ORDER BY` subclause, the list of candidate values is ordered in the same way as the underlying records and as specified by the associated `ORDER BY` subclause.
+
+In a regular aggregation (i.e. of the form `aggr(expr)`), the list of aggregated values is the list of candidate values with all `null` values removed from it.
+
+In a distinct aggregation (i.e. of the form `aggr(DISTINCT expr)`), the list of aggregated values is the list of candidate values with all `null` values removed from it.
+Futhermore, in a distinct aggregation, only the first of all equivalent candidate values is included in the list of aggregated values, i.e. duplicates under equivalence are removed.
```

Comment:
We should not say "only the first of all equivalent", we should say "only one of all equivalent", in order to not make any specific promises or requirements about _which_ value is picked. It _shouldn't_ matter, since they are equivalent, but let's not promise more than we need to in case it _could_ matter.